qiskit/transpiler/passes/mapping/algorithm/dependency_graph.py

```diff
@@ -0,0 +1,304 @@
+        """Name of the gate
+        Args:
+            i: Index of the gate in the `self._gates`
+        Returns:
+            Name of the gate.
+        """
+        return self._gates[i].name
+
+    def head_gates(self) -> FrozenSet[int]:
+        """Gates which can be applicable prior to the other gates
+        Returns:
+            Set of indices of the gates.
+        """
+        return frozenset([n for n in self._graph.nodes() if len(self._graph.in_edges(n)) == 0])
+
+    def gr_successors(self, i: int) -> List[int]:
+        """Successor gates in Gr (transitive reduction) of the gate
+        Args:
+            i: Index of the gate in the `self._gates`
+        Returns:
+            Set of indices of the successor gates.
+        """
+        return self._graph.successors(i)
+
+    def gate(self, gidx: int, layout: Layout, physical_qreg: QuantumRegister) -> InstructionContext:
+        """Convert acting qubits of gate `gidx` from virtual qubits to physical ones.
+        Args:
+            gidx: Index of the gate in the `self._gates`
+            layout: Layout used in conversion
+            physical_qreg: Register of physical qubit
+        Returns:
+            Converted gate with physical qubit.
+        Raises:
+            TranspilerError: if virtual qubit of the gate `gidx` is not found in the layout.
+        """
+        gate = copy.deepcopy(self._gates[gidx])
+        for i, virtual_qubit in enumerate(gate.qargs):
+            if virtual_qubit in layout.get_virtual_bits().keys():
+                gate.qargs[i] = Qubit(physical_qreg, layout[virtual_qubit])
+            else:
+                raise TranspilerError("virtual_qubit must be in layout")
+        return gate
+
+    def nx_graph(self) -> nx.DiGraph:
+        """Return deep copied networkx graph of this dependency graph.
+        """
+        return copy.deepcopy(self._graph)
+
+    @staticmethod
+    def remove_redundancy(graph):
+        """Remove redundant edges in DAG (= change `graph` to its transitive reduction)
+        """
+        edges = list(graph.edges())
+        for edge in edges:
+            graph.remove_edge(edge[0], edge[1])
+            if not nx.has_path(graph, edge[0], edge[1]):
+                graph.add_edge(edge[0], edge[1])
+
+    def _prior_gates_on_wire(self, toidx):
+        res = []
+        for qarg in self._gates[toidx].qargs:
+            res.append(reversed([i for i in self._gates_by_wire[qarg] if i < toidx]))
+        for carg in self._gates[toidx].cargs:
+            res.append(reversed([i for i in self._gates_by_wire[carg] if i < toidx]))
+        return res
+
+    def _create_gates_by_wire(self):
+        self._gates_by_wire = defaultdict(list)
+        for i, gate in enumerate(self._gates):
+            for qarg in gate.qargs:
+                self._gates_by_wire[qarg].append(i)
+            for carg in gate.cargs:
+                self._gates_by_wire[carg].append(i)
+
+
+if __name__ == '__main__':
```

Comment:
can this file be run stand-alone?

Reply:
even if it can, i don't think we should have this here.

Reply:
fixed at e341920